### PR TITLE
Copr: implement `RealDivide` for `RPN` function framework

### PR DESCRIPTION
**Pull request by iosmanthus**
# What have you changed? (mandatory) 
Add `RealDivide` to `RPN` function framework.

## What are the type of changes? (mandatory)
- New feature (a change which adds functionality)

## How has this PR been tested? (mandatory)
`cargo test impl_arithmetic` in `componets/tidb_query`

### CODE REVIEW

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 1]
```diff
@@ -1,5 +1,6 @@
 // Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
 
+use num_traits::identities::Zero;
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: why do we need this?

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 2]
```diff
@@ -510,6 +511,26 @@ impl ArithmeticOpWithCtx for DecimalDivide {
     }
 }
 
+pub struct RealDivide;
+
+impl ArithmeticOpWithCtx for RealDivide {
+    type T = Real;
+
+    fn calc(ctx: &mut EvalContext, lhs: &Real, rhs: &Real) -> Result<Option<Real>> {
+        Ok(if rhs.is_zero() {
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: rhs.into_inner() == 0?

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 3]
```diff
@@ -1253,4 +1274,90 @@ mod tests {
...
+                        ERR_DIVISION_BY_ZERO
+                    );
+                } else {
+                    assert!(ctx.take_warnings().warnings.is_empty());
+                }
+            }
+        }
+    }
+
+    fn double2real(x: f64) -> Option<Real> {
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: if you want to introduce this function, please also refine other places in this file to keep them consistent.

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 4]
```diff
@@ -1253,4 +1274,90 @@ mod tests {
...
+
+        for (lhs, rhs) in abnormal {
+            for &(flag, sql_mode, is_ok, has_warning) in &modes {
+                // Construct an `EvalContext`
+                let mut config = EvalConfig::new();
+                config.set_flag(flag).set_sql_mode(sql_mode);
+
+                let (result, mut ctx) = RpnFnScalarEvaluator::new()
+                    .context(EvalContext::new(std::sync::Arc::new(config)))
+                    .push_param(lhs.clone())
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: no need to clone a real?

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 5]
```diff
@@ -1253,4 +1274,90 @@ mod tests {
...
+
+            assert_eq!(actual, expected, "lhs={:?}, rhs={:?}", lhs, rhs);
+        }
+
+        let abnormal = vec![
+            (double2real(2.33), double2real(0.0)),
+            (double2real(2.33), double2real(-0.0)),
+        ];
+
+        // Vec<[(Flag, SqlMode, is_ok(bool), has_warning(bool))]>
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: Missing case:

```rust
            (
                ScalarFuncSig::DivideReal,
                Datum::F64(f64::MAX),
                Datum::F64(0.00001),
            ),
```

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 6]
```diff
@@ -1253,4 +1274,90 @@ mod tests {
...
+        let modes = vec![
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: These code looks very similar to `decimal_divide` and is very verbose. Please find a way to simplify it.

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs (hunk 2 above)
Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: Maybe `rhs.is_zero()` is more clear?

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs (hunk 1 above)
Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: `is_zero` is a method in trait `Zero`

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs (hunk 4 above)
Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: Fixed `clippy` warning

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs (hunk 5 above)
Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: OK.Marked

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs (hunk 6 above)
Comment (iosmanthus) [reply components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: OK. I will try.

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 7]
```diff
@@ -1189,15 +1194,69 @@ mod tests {
...
+    where
+        T: Evaluable + PartialEq,
+        Option<T>: Into<ScalarValue>,
+    {
         // Vec<[(Flag, SqlMode, is_ok(bool), has_warning(bool))]>
         let modes = vec![
             // Warning
             (Flag::empty(), SqlMode::empty(), true, true),
-            // Error
+            // Erroxur
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: typo

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 8]
```diff
@@ -1220,17 +1279,17 @@ mod tests {
...
 
-        for (lhs, rhs) in abnormal {
+        for args in cases {
             for &(flag, sql_mode, is_ok, has_warning) in &modes {
                 // Construct an `EvalContext`
                 let mut config = EvalConfig::new();
                 config.set_flag(flag).set_sql_mode(sql_mode);
 
-                let (result, mut ctx) = RpnFnScalarEvaluator::new()
+                let evaluator_builder = RpnFnScalarEvaluator::new()
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: why not `RpnFnScalarEvaluator::new().context(..).push_params(..).evaluate_ctx::<T>(..)`?

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 9]
```diff
@@ -1189,15 +1194,69 @@ mod tests {
...
 
+        test_divide_by_zero(divide_by_zero, ScalarFuncSig::DivideReal);
+    }
+
+    fn double2real(x: f64) -> Option<Real> {
+        Real::new(x).ok()
+    }
+
+    // Check if `ctx` is correct in several modes with cases
+    fn test_divide_by_zero<T>(cases: Vec<Vec<Option<T>>>, sig: ScalarFuncSig)
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: Looks like there is no need to use generics:

```rust
#[test]
fn test_divide_by_zero() {
    let cases = ...;
    for ... {
    }
}
```

is more readable.

File: components/tidb_query/src/rpn_expr/impl_arithmetic.rs [hunk 10]
```diff
@@ -1189,15 +1194,69 @@ mod tests {
...
+
+        let divide_by_zero = vec![
+            vec![double2real(2.33), double2real(0.0)],
+            vec![double2real(2.33), double2real(-0.0)],
         ];
 
+        test_divide_by_zero(divide_by_zero, ScalarFuncSig::DivideReal);
+    }
+
+    fn double2real(x: f64) -> Option<Real> {
```
Comment (breeswish) [inline components/tidb_query/src/rpn_expr/impl_arithmetic.rs]: No need to convert double to real in order to feed into `push_param`:

```rust
impl From<Option<f64>> for ScalarValue {
    #[inline]
    fn from(s: Option<f64>) -> ScalarValue {
        ScalarValue::Real(s.and_then(|f| Real::new(f).ok()))
    }
}

impl From<f64> for ScalarValue {
    #[inline]
    fn from(s: f64) -> ScalarValue {
        ScalarValue::Real(Real::new(s).ok())
    }
}
```

**Pull request merged**